// @antchain/myassembly/assembly/contracts/DRC-721/standard.ts
// DRC-721 规范
// https://yuque.antfin-inc.com/docs/share/879ddb35-fe73-4b72-aaa3-a6424f330ab5

// 资产的元信息
@unmanaged
export class DRC721_AssetMeta {
  // 发布者的账户地址
  issuer: string;
  // 资产描述，通常描述了资产属性与规则等
  description: string | null;
  // 资产链接的前面部分
  baseUri: string | null;
}

// 资产信息
@unmanaged
export class DRC721_DigitalAsset {
  // 资产编号，在当前合约中唯一，通常是自增 ID
  assetId: string;
  // 资产信息，通常是属性值或哈希值，比较复杂的也可以用 JSON 字符串表达
  assetInfo: string | null;
  // 资产链接，通常是多媒体资源的 IPFS 或 CDN 地址
  assetUri: string | null;
}

export abstract class IDRC721 {
  /**
   * 获取当前资产的元信息
   * @returns AssetMeta JSON string
   */
  abstract AssetMeta(): string;

  /**
   * 获取指定资产的信息
   * @param assetId 资产 ID
   * @returns 资产信息 DigitalAsset JSON string
   */
  abstract GetAsset(assetId: string): string;

  /**
   * 查询指定资产的拥有者
   * @param assetId 要查询的资产 ID
   * @returns string 拥有者的账户地址
   */
  abstract OwnerOf(assetId: string): string;

  /**
   * 发放资产
   * @param owner 发放的账户地址
   * @param asset 资产信息 DRC721_AssetDetail JSON string
   */
  abstract Issue(owner: string, assetStr: string): void;


  // /**
  //  * 替换资产信息
  //  * @param owner 
  //  * @param assertStr 
  //  */
  // abstract ReplaceIssue(owner:string,assertStr:string):void;

  

  /**
   * 转移资产
   * @param to 要转移到的账户地址
   * @param assetId 要转移的资产 ID
   */
  abstract Transfer(to: string, assetId: string): void;

  /**
   * 查看指定账户是否拥有指定资产
   * @param owner 要查询的账户地址
   * @param assetId 要查询的资产 ID
   * @returns boolean
   */
  abstract HasAsset(owner: string, assetId: string): bool;

  /**
   * 查询指定账户拥有的资产数量
   * @param owner 要查询的账户地址
   * @returns u64 资产数量
   */
  abstract AssetCount(owner: string): u64;
}
